xds/internal/balancer/outlierdetection/balancer.go

```diff
@@ -105,3 +132,785 @@ func (bb) ParseConfig(s json.RawMessage) (serviceconfig.LoadBalancingConfig, err
+
+}
+
+// syccessRateAlgorithm ejects any addresses where the success rate falls below
+// the other addresses according to mean and standard deviation, and if overall
+// applicable from other set heuristics.
+func (b *outlierDetectionBalancer) successRateAlgorithm() {
+	// 1. If the number of addresses with request volume of at least
+	// success_rate_ejection.request_volume is less than
+	// success_rate_ejection.minimum_hosts, stop.
+	if b.numAddrsWithAtLeastRequestVolume() < b.cfg.SuccessRateEjection.MinimumHosts { // TODO: O(n) search, is there a way to optimize this?
+		return
+	}
+
+	// 2. Calculate the mean and standard deviation of the fractions of
+	// successful requests among addresses with total request volume of at least
+	// success_rate_ejection.request_volume.
+	mean, stddev := b.meanAndStdDevOfSuccessesAtLeastRequestVolume()
+
+	// 3. For each address:
+	for _, addrInfo := range b.addrs {
+		ccb := addrInfo.callCounter.inactiveBucket
+		sre := b.cfg.SuccessRateEjection
+		// i. If the percentage of ejected addresses is greater than
+		// max_ejection_percent, stop.
+		if float64(b.numAddrsEjected)/float64(len(b.addrs))*100 > float64(b.cfg.MaxEjectionPercent) {
+			return
+		}
+
+		// ii. If the address's total request volume is less than
+		// success_rate_ejection.request_volume, continue to the next address.
+		if ccb.requestVolume < int64(sre.RequestVolume) {
+			continue
+		}
+
+		//  iii. If the address's success rate is less than (mean - stdev *
+		//  (success_rate_ejection.stdev_factor / 1000))
+		successRate := float64(ccb.numSuccesses) / float64(ccb.requestVolume)
+		if successRate < (mean - stddev*(float64(sre.StdevFactor)/1000)) {
+			// then choose a random integer in [0, 100). If that number is less
+			// than success_rate_ejection.enforcement_percentage, eject that
+			// address.
+			if uint32(grpcrand.Int31n(100)) < sre.EnforcementPercentage {
+				b.ejectAddress(addrInfo)
+			}
+		}
+	}
+}
+
+// failurePercentageAlgorithm ejects any addresses where the failure percentage
+// rate exceeds a set enforcement percentage, if overall applicable from other
+// set heuristics.
+func (b *outlierDetectionBalancer) failurePercentageAlgorithm() {
+	// 1. If the number of addresses is less than
+	// failure_percentage_ejection.minimum_hosts, stop.
+	if uint32(len(b.addrs)) < b.cfg.FailurePercentageEjection.MinimumHosts {
+		return
+	}
+
+	// 2. For each address:
+	for _, addrInfo := range b.addrs {
+		ccb := addrInfo.callCounter.inactiveBucket
+		fpe := b.cfg.FailurePercentageEjection
+		// i. If the percentage of ejected addresses is greater than
+		// max_ejection_percent, stop.
+		if float64(b.numAddrsEjected)/float64(len(b.addrs))*100 > float64(b.cfg.MaxEjectionPercent) {
+			return
+		}
+		// ii. If the address's total request volume is less than
+		// failure_percentage_ejection.request_volume, continue to the next
+		// address.
+		if uint32(ccb.requestVolume) < fpe.RequestVolume {
```

Comment:
Cast from int64->uint32 -- is this safe?  Can `requestVolume` be a uint32 instead if it is safe?

Reply:
Ah, great point. Since the configuration defines integers as uint32, capping the configuration to 32 bits of data, and since these numbers can't be negative, I switched every field in bucket to uint32.